library/src/main/java/com/google/android/exoplayer2/extractor/mp4/AtomParsers.java

```diff
@@ -423,47 +430,333 @@ private static void parseMetaAtom(ParsableByteArray data, GaplessInfoHolder out)
+    int length = ilst.readInt() - Atom.FULL_HEADER_SIZE;
+    int key = ilst.readInt();
+    ilst.skipBytes(4);
+    if (key == Atom.TYPE_data) {
+      Object value = parseDataBox(ilst, length);
+      if (value instanceof byte[]) {
+        byte[] bytes = (byte[]) value;
+        if (bytes.length == 2) {
+          int code = (bytes[0] << 8) + (bytes[1] & 0xFF);
+          String s = Id3Decoder.decodeGenre(code);
+          if (s != null) {
+            Id3Frame frame = new TextInformationFrame(attributeName, s);
+            builder.add(frame);
           }
         }
-        if (lastCommentName != null && lastCommentData != null
-            && "com.apple.iTunes".equals(lastCommentMean)) {
-          out.setFromComment(lastCommentName, lastCommentData);
-          break;
+      }
+    } else {
+      ilst.skipBytes(length);
+    }
+  }
+
+  private static void parseExtendedAttribute(MetadataBuilder builder,
+                                             ParsableByteArray ilst,
+                                             int endPosition) {
+    String domain = null;
+    String name = null;
+    Object value = null;
+
+    while (ilst.getPosition() < endPosition) {
+      int length = ilst.readInt() - Atom.FULL_HEADER_SIZE;
+      int key = ilst.readInt();
+      ilst.skipBytes(4);
+      if (key == Atom.TYPE_mean) {
+        domain = ilst.readNullTerminatedString(length);
+      } else if (key == Atom.TYPE_name) {
+        name = ilst.readNullTerminatedString(length);
+      } else if (key == Atom.TYPE_data) {
+        value = parseDataBox(ilst, length);
+      } else {
+        ilst.skipBytes(length);
+      }
+    }
+
+    if (value != null) {
+      if (Util.areEqual(domain, "com.apple.iTunes") && Util.areEqual(name, "iTunSMPB")) {
+        String s = value instanceof byte[] ? new String((byte[]) value) : value.toString();
+        builder.setGaplessInfo(GaplessInfo.createFromComment("iTunSMPB", s));
+      }
+
+      if (Util.areEqual(domain, "com.apple.iTunes") && Util.areEqual(name, "iTunNORM") && (value instanceof byte[])) {
+        String s = new String((byte[]) value);
+        Id3Frame frame = new CommentFrame("eng", "iTunNORM", s);
+        builder.add(frame);
+      } else if (domain != null && name != null) {
+        String extendedName = domain + "." + name;
+        if (value instanceof String) {
+          Id3Frame frame = new TextInformationFrame(extendedName, (String) value);
+          builder.add(frame);
+        } else if (value instanceof Integer) {
+          Id3Frame frame = new TextInformationFrame(extendedName, value.toString());
+          builder.add(frame);
+        } else if (value instanceof byte[]) {
+          byte[] bb = (byte[]) value;
+          Id3Frame frame = new BinaryFrame(extendedName, bb);
+          builder.add(frame);
         }
+      }
+    }
+  }
+
+  private static Object parseDataBox(ParsableByteArray ilst, int length) {
+    int versionAndFlags = ilst.readInt();
```

Comment:
I think this ends up reading the wrong 4 bytes. At the calling sites it appears you skip over the versionAndFlags int. This the reads the following 4 bytes, which I think are defined to be empty bytes in a data box.